src/core/tests/type_prop/segment_max.cpp

```diff
@@ -0,0 +1,310 @@
+
+    const auto op = make_op(data, segment_ids, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 12, 81}));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, all_inputs_dynamic) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape::dynamic());
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape::dynamic());
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape::dynamic());
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape::dynamic()));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, all_inputs_interval) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape{{2, 4}, {12, 15}, {18, 300}});
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape{{2, 4}});
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape::dynamic());
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), {12, 15}, {18, 300}}));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, data_dynamic) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape::dynamic());
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape{6});
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape{});
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape::dynamic()));
+    EXPECT_EQ(op->get_fill_mode(), op::FillMode::ZERO);
+}
+
+TEST_F(TypePropSegmentMaxTest, segment_ids_dynamic) {
+    const auto data = std::make_shared<Parameter>(element::f32, PartialShape{3, 12, 81});
+    const auto segment_ids = std::make_shared<Parameter>(element::i64, PartialShape::dynamic());
+    const auto num_segments = std::make_shared<Parameter>(element::i64, PartialShape{});
+
+    const auto op = make_op(data, segment_ids, num_segments, op::FillMode::ZERO);
+    op->validate_and_infer_types();
+
+    EXPECT_EQ(op->get_output_element_type(0), element::f32);
+    EXPECT_EQ(op->get_output_partial_shape(0), (PartialShape{Dimension::dynamic(), 12, 81}));
```

Comment:
The dynamism of the first dimension of the output is more related to the `num_segments` as Parameter then the dynamism of `segment_ids`.
Let's add similar case but without the `num_segment` input provided to the op constructor.

Reply:
Added multiple combinations of such tests, it'll be easier to see them by inspecting the latest commit.